### ISSUE TITLE
Target igl::all

### ISSUE THREAD

**Issue by teseoch** [libigl/libigl#1551]
#### Describe your feature request

igl provides multiple targets depending on which options are on/off
usually one wants to enable all necessary ones and link against all targets.

I normally find myself missing one (or look at the cmake to know the name).
It would be nice to have an (interface?) target `igl::all` that includes all active targets 

#### Check all that apply (change to `[x]`)
- [x] Windows
- [x] macOS
- [x] Linux

**Comment by alecjacobson** [comment]
this is a good idea. Would `igl::all` mean _all that are turned on_ or really _all_?

**Comment by jdumas** [comment]
I don't think it's a good idea. The meaning of the target `igl::all` will change depending on which CMake option you have enabled, which means that a project linking against `igl::all` may or may not work depending on which option are enabled by the client (e.g. imagine you are writing a lib like polyfem, which depends on `igl::all`). This is not a good way to express dependencies, and should be discouraged. A strength of having to explicitly `igl::opengl`, or `igl::png` are CMake time is that you see immediately that you have a missing dependency and that you should enabled the corresponding CMake option.

Maybe we should look to improve our current target discoverability instead. E.g. how do you determine that you need a new dependency? The CMake target name should follow from the include path directly.

**Comment by teseoch** [comment]
My problem is the opposite. I have the list of options, then I never remember which target to link against.

I don't understand the might or might not work. In general this is true if you have a project with some options on if you randomly change them it might not work anymore.
I partially agree with the polyfem example, but this is rarer.
The standard usage of libigl is more for applications than developing libraries, which is where my comment is coming from.

What I have in mind when you create a new cmake project is
set some options on
link against `igl::all`

if later you need other options on, change them and you are good to go.

**Comment by jdumas** [comment]
I would rather have an option `LIBIGL_ALL` that define all targets and then you would have to explicitly list dependencies such as `igl::core`, `igl::png`, etc. If you have defined the option `LIBIGL_WITH_TETGEN` then the target name should be automatically deduced as `igl::tetgen` (remove the LIB, the WITH, and make it lowercase). So again if you have the list of options you have the list of targets no?

The scenario I'm trying to avoid is the following:
- libA is using `igl::core` and `igl::triangle`
- libB is using `igl::core` and `igl::embree`
- main app is using both libA and libB, thus it needs to enable both `LIBIGL_WITH_EMBREE` and `LIBIGL_WITH_TRIANGLE` in the root CMake

Now, if both libA and libB are linking against `igl::all`, then because of the main app, suddenly libA is linking against embree. This behavior can cause problems for a number of reason (symbol collisions, file size, header pollution, etc.). Thus it should be discouraged. On the other hand, if the main app doesn't define the necessary CMake options, then either libA or libB will not compile, and you will only find out after you actually *try* to compile then, not at CMake time.

> The standard usage of libigl is more for applications than developing libraries, which is where my comment is coming from.

That may be true for the typical usage in a research project, but there is also a lot of library (either open-source or not) that depend on libigl. Not saying they're all using our CMake, but still...

**Comment by teseoch** [comment]
I agree with your example, but in that case, libA and libB need to do something like `if(TARGET igl::core) ...` which implies that the developers of libA and libB know what they are doing.

I am not saying to remove the current targets, but add a new one to facilitate end-applications of libigl

**Comment by jdumas** [comment]
> in that case, libA and libB need to do something like if(TARGET igl::core) ...

What do you mean? The typical usage should be as follows:

```cmake
# libA/CMakeList.txt
option(LIBIGL_WITH_TRIANGLE "" ON)
find_package(libigl) # or add_subdirectory()
target_link_library(libA PUBLIC igl::core igl::triangle)

# libA/CMakeList.txt
option(LIBIGL_WITH_EMBREE "" ON)
find_package(libigl) # or add_subdirectory()
target_link_library(libB PUBLIC igl::core igl::embree)

# CMakeLists.txt
option(LIBIGL_WITH_TRIANGLE "" ON)
option(LIBIGL_WITH_EMBREE "" ON)
add_subdirectory(libA)
add_subdirectory(libB)
target_link_library(app PUBLIC libA libB)
```

> I am not saying to remove the current targets, but add a new one to facilitate end-applications of libigl

I think it encourages a bad practice for the sake of a (minor) convenience. With the amount of people out there that have no idea how to use CMake, I'm not sure we want to do that. And again why is the current situation not practical already? You already know you that if you need the option `LIBIGL_WITH_EMBREE`, then the target you want is `igl::embree`.
